DOCSP-38031 Updates to Authentication Provider Metadata (#714)
mongodb/docs-app-services: source/users/custom-metadata.txt

- **Custom Data May Be Stale**: A user's custom data is *sourced* from a
  MongoDB collection but is stored in and read from a user's
  authentication access token. If a user's metadata is updated after the 
  access token is issued, requests using the previously created access 
  token will not have the newly updated metadata. Their user metadata 

>   will update when they refresh their access token or re-authenticate.

Reviewer: [Custom user data](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/#custom-user-data) and [user metadata from authentication providers](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/#authentication-provider-metadata) are two separate things so we would _not_ want to overwrite this. Instead we would want to just add this section you wrote most probably to [this section](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/#authentication-provider-metadata)

       <custom-jwt-authentication-configuration-metadata-fields>`
       configuration.

.. important:: Security and Authentication Provider Metadata 

>   Authentication provider metadata is externally-defined by clients and

Reviewer: [nit] 
```suggestion
  Authentication provider metadata can be externally-defined by clients and
```

  external authentication providers and should be regarded with caution. You should 
  not rely solely on authentication provider metadata for security-related 
  decisions, such as using this metadata in rule expressions for data access 

>   permssions.

Reviewer: [typo]
```suggestion
  permissions.
```

                          "name": "jwt.field.path",
                          "field_name": "metadataFieldName"
                        }
                      ],
                      "config": {
                        "audience": [],
                        "requireAnyAudience": false,
                        "signingAlgorithm": "HS256",
                        "useJWKURI": false
                      },
                      "secret_config": {
                        "signingKeys": [
                          "<JWT Signing Key>"
                        ]
                      },
                      "disabled": true
                      }'

          If you successfully configure the provider's metadata
          fields, App Services returns a ``204`` response.

.. note:: Tips for Users with Multiple Linked Authentication Providers  

  - To ensure the most up-to-date metadata, users should re-authenticate after 
    switching to a different authentication provider. Failure to do so may result 
    in outdated metadata reflected in the :guilabel:`Users` table in 
    the UI's :guilabel:`App Users` page and for requests using authentication providers.

>   - If a user switches between authentication providers, it may take up to 30 minutes 
>     to propagate metadata to the :guilabel:`Users` table. The request is always guaranteed to have the 

Reviewer: after the comment above, can we just actually remove the "Users table" bit? Since the above bullet has both "Users Table" and "requests using authentication methods", I don't want this next bullet to only mention one of those (when it does apply to both but we're not including it). 
```suggestion
  - If a user switches between authentication providers, it may take up to 30 minutes 
   for metadata to propagate. The request is always guaranteed to have the 
```

         If you successfully configure the provider's metadata
         fields, App Services returns a ``204`` response.

.. note:: Tips for Users with Multiple Linked Authentication Providers  

  - To ensure the most up-to-date metadata, users should re-authenticate after 
    switching to a different authentication provider. Failure to do so may result 

>     in outdated metadata reflected in the user's table and for requests 

Reviewer: This is a table with a list of all the users so I don't think this should be possessive!
```suggestion
    in outdated metadata reflected in the Users Table and for requests 
```

Author: Do you know whether the table name is ``user`` or ``users``? Would like to rephrase user table to 'user collection', as we don't typically refer to data as being stored in tables in the docs

Reviewer: ah when I say "Users Table", I mean an actual table component in our UI that contains a list of users for the admin. But actually, it might be clearer to say the "Users Page" instead if you're good with that!

Author: ok makes sense, thank you! Updated the wording, let me know what you think!

    using authentication providers.

>   - If a user switches between authentication providers, there may be a 30 minute delay 
>     to update metadata that has the same field name for both authentication 
>     providers. For example, both JWT and API Key authentication providers have 
>     a ``name`` metadata field. There will be a delay to update the ``name`` field 
>     in the user's table if the user authenticates with JWT and then with an API key. 
>     The request is always guaranteed to have the most up-to-date metadata associated 
>     with the authentication provider used.

Reviewer: I don't think this example is necessary since the case is an edge case and there'd have to be other factors involved for this data to become stale--we don't want to confuse users unnecessarily. Also JWT authentication isn't guaranteed to have a "name" field. 

This also makes it seem like we'll only ever have stale metadata regarding the conflicting field name when in fact, other metadata could possibly be stale (especially if we end up expanding on our caching in the future).
```suggestion
  - If a user switches between authentication providers, there may be a 30 minute delay in propagating metadata to the Users Table and requests using authentication providers. The request is always guaranteed to have the most up-to-date metadata associated with the current authentication provider used.
```

         fields, App Services returns a ``204`` response.

.. note:: Tips for Users with Multiple Linked Authentication Providers  

  - To ensure the most up-to-date metadata, users should re-authenticate after 
    switching to a different authentication provider. Failure to do so may result 
    in outdated metadata reflected in the :guilabel:`Users` table in 
    the UI's :guilabel:`App Users` page and for requests using authentication providers.

>   - If a user switches between authentication providers, there may be a 30 minute delay 
>     in propagating metadata to the :guilabel:`Users` table and for requests using 

Reviewer 2: Minor suggestion for concision and avoid internal implementation detail that's not relevant to users.
```suggestion
  - If a user switches between authentication providers, it may take 30 minutes to
     to propagate metadata.
```

It's not clear to me how "and for requests using authentication providers" relates. Is there also a 30 minute delay in metadata there? If so, the next sentence seems to contradict that.

Author: Removed the contradicting bit! Left the Users table part, as the engineer wanted to highlight that detail specifically to users.

Reviewer: Yeah so the original message we were trying to convey was, "Each request is guaranteed to have the most up to date metadata associated with their auth provider, however changes can take up to 30 minutes to propagate to the users table and requests using other authentication providers." just to make it clear that authenticating with a _different_ authentication provider may not have the metadata specified in the request made prior with the original authentication provider used. 

With how this PR currently words this, I'd be good with going with @krollins-mdb suggestion and removing the "Users Table" part out of this as well!
